### ISSUE TITLE
Function redefinition is flagged if the types are generic

### ISSUE THREAD

**Issue by nedbat** [microsoft/pyright#11729]
**Describe the bug**
Conditionally defining a function two different ways is reported as a redeclaration if the types are generic, but not if the types are definite.

**Code or Screenshots**

```python
from collections.abc import Callable
from typing import ParamSpec, TypeVar

P = ParamSpec("P")
R = TypeVar("R")

try:
    import coverage  # pyright: ignore[reportUnusedImport]
except ImportError:

    def f(a: str) -> int:
        return 1

    def fgen(func: Callable[P, R]) -> Callable[P, R]:
        return func
else:

    def f(a: str) -> int:
        return 2

    def fgen(func: Callable[P, R]) -> Callable[P, R]:
        return func
```

Running with 1.1.413, I get:
```
% pyright /tmp/bug.py
/tmp/bug.py
  /tmp/bug.py:21:9 - error: Function declaration "fgen" is obscured by a declaration of the same name (reportRedeclaration)
1 error, 0 warnings, 0 informations
```

`f` is fine to redefine, but `fgen` is considered a problem.  Both are the same: the name has the same type under either condition.
